### PR TITLE
Skew coefficient

### PR DESCRIPTION
**Pull request by ghosalsattam**
Add Skew Coefficient to projection Matrix. See #335 .

### CODE REVIEW

**Comment by ghosalsattam** [comment]
@julienvalentin please provide a review.

**Comment by ghosalsattam** [comment]
@julienvalentin Do you suggest any more changes?

**Comment by julienvalentin** [comment]
Thanks for the PR @ghosalsattam !

In vacation at the moment, re-assigning.

**Comment by ghosalsattam** [comment]
@sofienbouaziz Please provide a review.

**Comment by ghosalsattam** [comment]
@taiya @sofienbouaziz @julienvalentin Do I need to make more changes?

**Comment by fcole** [comment]
Sorry for the delay, I can take a look at this. Mostly it looks fine, but there are many whitespace changes (changed tab widths?). Could you remove the whitespace changes to narrow the change to the code you added?

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 1]
```diff
@@ -201,40 +224,43 @@ def intrinsics_from_matrix(matrix, name=None):
       "perspective_intrinsics_from_matrix".
 
   Returns:
-    Tuple of two tensors, each one of shape `[A1, ..., An, 2]`. The first
-    tensor represents the focal length, and the second one the principle point.
+    Tuple of three tensors, first two of shape `[A1, ..., An, 2]` and the third of shape `[A1, ..., An, 1]`. The first
+    tensor represents the focal length, and the second one the principle point. The third represent Skew Coefficient.
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:224]: nit on docstring: how about "The first tensor represents the focal length, the second the principal point, and the third the skew."

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 2]
```diff
@@ -201,40 +224,43 @@ def intrinsics_from_matrix(matrix, name=None):
...
+          tensor_name="matrix",
+          has_rank_greater_than=1,
+          has_dim_equals=((-1, 3), (-2, 3)),
+      )
+
+      fx = matrix[..., 0, 0]
+      fy = matrix[..., 1, 1]
+      cx = matrix[..., 0, 2]
+      cy = matrix[..., 1, 2]
+      skew_coefficient = matrix[..., 0, 1]
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:243]: I'd just call this "skew" instead of "skew_coefficient", since I feel the "coefficient" is implied (here and elsewhere in this file).

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 3]
```diff
@@ -262,34 +290,42 @@ def matrix_from_intrinsics(focal, principal_point, name=None):
...
+          tensor_name="principal_point",
+          has_dim_equals=(-1, 2),
+      )
+      shape.compare_batch_dimensions(
+          tensors=(focal, principal_point),
+          tensor_names=("focal", "principal_point"),
+          last_axes=-2,
+          broadcast_compatible=False,
+      )
+      skew_coefficient_shape = skew_coefficient.shape.as_list()
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:308]: Please use similar shape checking code to the principal_point and focal above, and add skew to the compare_batch_dimensions call. Alternatively, use the get_broadcasted_shape utility to broadcast each tensor to a common shape, as in for example https://github.com/tensorflow/graphics/blob/3dd63357f54f42ad60c6dc0daeb3e1739f7863be/tensorflow_graphics/geometry/transformation/rotation_matrix_3d.py#L398

File: tensorflow_graphics/rendering/camera/perspective.py (hunk 2 above)
Comment (ghosalsattam) [reply tensorflow_graphics/rendering/camera/perspective.py]: Done

**Comment by ghosalsattam** [comment]
@fcole I have addressed all your suggested review.

**Comment by fulkast** [comment]
Hey @fcole @ghosalsattam! Great work on this issue!

I opened a separate PR #420 not noticing that this PR was already in flight. There are some additions in #420 that are not present here i.e adding the skew parameter to the `projection/unprojection` functions. May I rebase off of this branch and have my additions reviewed in a separate PR?

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 4]
```diff
@@ -266,27 +273,36 @@ def matrix_from_intrinsics(focal, principal_point, name=None):
...
         tensor_name="principal_point",
-        has_dim_equals=(-1, 2))
+        has_dim_equals=(-1, 2),
+    )
+    shape.check_static(
+        tensor=skew,
+        tensor_name="skew",
+        has_dim_equals=(-1, 1),
+    )
     shape.compare_batch_dimensions(
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:289]: Sorry for not catching this earlier, but I think setting a default value of (0.0,) for skew and then raising an error through `compare_batch_dimensions` if the shapes don't match will break existing code that passes additional batch dimensions for `focal` and `principal_point` (right?). I think this function needs to use the `broadcast_to` style of https://github.com/tensorflow/graphics/blob/3dd63357f54f42ad60c6dc0daeb3e1739f7863be/tensorflow_graphics/geometry/transformation/rotation_matrix_3d.py#L398
This behavior really should be treated more consistently in tensorflow_graphics as a whole, but that's another issue...

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 5]
```diff
@@ -266,27 +273,36 @@ def matrix_from_intrinsics(focal, principal_point, name=None):
...
     zero = tf.zeros_like(fx)
     one = tf.ones_like(fx)
-    matrix = tf.stack((fx, zero, cx,
-                       zero, fy, cy,
-                       zero, zero, one),
-                      axis=-1)  # pyformat: disable
+
+    skew = tf.reshape(skew, tf.shape(fx))
+
+    matrix = tf.stack(
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:303]: There are still unnecessary whitespace changes in this file (for example this matrix). Please remove all unnecessary whitespace changes.

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 6]
```diff
@@ -201,40 +201,45 @@ def intrinsics_from_matrix(matrix, name=None):
...
     focal = tf.stack((fx, fy), axis=-1)
     principal_point = tf.stack((cx, cy), axis=-1)
-  return focal, principal_point
+    skew = tf.expand_dims(skew, axis=0)
 
+    return focal, principal_point, skew
 
-def matrix_from_intrinsics(focal, principal_point, name=None):
+
+def matrix_from_intrinsics(focal, principal_point, skew=(0.0,), name=None):
```
Comment (fcole) [inline tensorflow_graphics/rendering/camera/perspective.py:234]: Given that the shape of skew has to be coerced to match focal and prinicipal_point anyway, it probably makes sense for the default value to be just the scalar "0.0", not the tuple "(0.0,)".

**Comment by fcole** [comment]
> Hey @fcole @ghosalsattam! Great work on this issue!
> 
> I opened a separate PR #420 not noticing that this PR was already in flight. There are some additions in #420 that are not present here i.e adding the skew parameter to the `projection/unprojection` functions. May I rebase off of this branch and have my additions reviewed in a separate PR?

Hi @fulkast, sounds good, thanks! I took a look at your PR and the differences also seem like good changes. I left a comment regarding broadcast compatibility in #420.

File: tensorflow_graphics/rendering/camera/perspective.py (hunk 4 above)
Comment (ghosalsattam) [reply tensorflow_graphics/rendering/camera/perspective.py:289]: @fcole Will you please elaborate?

File: tensorflow_graphics/rendering/camera/tests/perspective_test.py [hunk 7]
```diff
@@ -239,18 +239,22 @@ def test_intrinsics_from_matrix_to_intrinsics_random(self):
...
+    random_skew_coeff = np.reshape(random_skew_coeff, (1, 1))
 
     self.assertAllClose(random_focal, focal, rtol=1e-3)
     self.assertAllClose(random_principal_point, principal_point, rtol=1e-3)
+    self.assertAllClose(random_skew_coeff, skew_coeff, rtol=1e-3)
 
   @parameterized.parameters(
-      ((2,), (2,)),
-      ((2, 2), (2, 2)),
-      ((None, 2), (None, 2)),
```
Comment (fulkast) [inline tensorflow_graphics/rendering/camera/tests/perspective_test.py:253]: This original test for instance would fail, if the batch dimension of `skew` is not coerced before `compare_batch_dimensions` is called.

File: tensorflow_graphics/rendering/camera/tests/perspective_test.py [hunk 8]
```diff
@@ -239,18 +239,22 @@ def test_intrinsics_from_matrix_to_intrinsics_random(self):
...
+    focal, principal_point, skew_coeff = perspective.intrinsics_from_matrix(matrix)
+    random_skew_coeff = np.reshape(random_skew_coeff, (1, 1))
 
     self.assertAllClose(random_focal, focal, rtol=1e-3)
     self.assertAllClose(random_principal_point, principal_point, rtol=1e-3)
+    self.assertAllClose(random_skew_coeff, skew_coeff, rtol=1e-3)
 
   @parameterized.parameters(
-      ((2,), (2,)),
-      ((2, 2), (2, 2)),
```
Comment (fulkast) [inline tensorflow_graphics/rendering/camera/tests/perspective_test.py:252]: This original test for instance would fail, if the batch dimension of `skew` is not coerced before `compare_batch_dimensions` is called.

File: tensorflow_graphics/rendering/camera/perspective.py (hunk 4 above)
Comment (fulkast) [reply tensorflow_graphics/rendering/camera/perspective.py:289]: @ghosalsattam what happens if you didn't change the test parameters on this line here: https://github.com/tensorflow/graphics/pull/406/files#r459409028 ? Is there an error?

File: tensorflow_graphics/rendering/camera/tests/perspective_test.py (hunk 8 above)
Comment (fcole) [reply tensorflow_graphics/rendering/camera/tests/perspective_test.py:252]: After some discussion with @sofienbouaziz , it may make the most sense to actually just remove the shape checking for this matrix_from_intrinsics. It seems like the function will work as is ok if skew=0.0, since it's reshaped to the same shape as fx, so the shape checking is causing more problems than its solving in this case, I believe.

File: tensorflow_graphics/rendering/camera/perspective.py (hunk 4 above)
Comment (ghosalsattam) [reply tensorflow_graphics/rendering/camera/perspective.py:289]: @fcole @fulkast What should be done now? Shall we leave the code at the current state in this regard or add broadcast_to/remove compare_batch_dimensions or shall we add broadcast_to?

Comment (fcole) [reply tensorflow_graphics/rendering/camera/perspective.py:289]: Sorry for the delay. Could you check what happens if you leave `skew=0.0` as the default argument, remove the shape checking code entirely and try the following:

1. run the original test with no skew argument (to simulate backwards compatibility)
2. run the new test with skew set. 

If that all passes ok, then I lean towards just removing the checking code.

**Comment by ghosalsattam** [comment]
@fcole I had run the code which was giving the error when shape checking was removed. So I added the broadcast_to method and now all the tests are running(both passing skew nd not passing skew).

**Comment by fcole** [comment]
Cool, thanks, this is looking pretty good. Please clean up the unnecessary whitespace changes and make sure the tests pass (the failure seems unrelated to this PR... maybe you need to sync to head?)

**Comment by ghosalsattam** [comment]
@fcole done!

**Comment by ghosalsattam** [comment]
@fcole Please provide a review.

**Comment by ghosalsattam** [comment]
@fcole I have done the changes.

**Comment by fcole** [comment]
@sofienbouaziz please take a look

**Comment by ghosalsattam** [comment]
@fcole are there any rfcs for this repository(tensorflow graphics).
If yes can you provide me a link?

File: tensorflow_graphics/rendering/camera/perspective.py [hunk 9]
```diff
@@ -201,40 +201,42 @@ def intrinsics_from_matrix(matrix, name=None):
...
-
     fx = matrix[..., 0, 0]
     fy = matrix[..., 1, 1]
     cx = matrix[..., 0, 2]
     cy = matrix[..., 1, 2]
+    skew = matrix[..., 0, 1]
     focal = tf.stack((fx, fy), axis=-1)
     principal_point = tf.stack((cx, cy), axis=-1)
-  return focal, principal_point
+    skew = tf.expand_dims(skew, axis=0)
```
Comment (fulkast) [inline tensorflow_graphics/rendering/camera/perspective.py:227]: Hey @ghosalsattam @fcole 

I hope you all had a great weekend

Sorry I didn't catch this earlier. But wouldn't this line create a tensor of shape `[1, A1, ... An]` instead of the desired `[A1, ... An, 1]`?

Shouldn't it be `skew = tf.expand_dims(skew, axis=-1)` instead?

Comment (ghosalsattam) [reply tensorflow_graphics/rendering/camera/perspective.py]: @fulkast, Thanks for pointing out and sorry for delayed response.
I have corrected this part.

**Review by G4G — APPROVED** [review]

**Review by G4G — APPROVED** [review]
